lib/web/app/transport.go

```diff
@@ -231,41 +225,36 @@ func (t *transport) rewriteRequest(r *http.Request) error {
 
 // DialContext dials and connect to the application service over the reverse
 // tunnel subsystem.
-func (t *transport) DialContext(ctx context.Context, _, _ string) (net.Conn, error) {
-	var err error
-	var conn net.Conn
-
-	t.servers.Range(func(serverID, appServerInterface interface{}) bool {
-		appServer, ok := appServerInterface.(types.AppServer)
-		if !ok {
-			t.c.log.Warnf("Failed to load AppServer, invalid type %T", appServerInterface)
-			return true
-		}
-
-		var dialErr error
-		conn, dialErr = dialAppServer(ctx, t.c.proxyClient, t.c.identity.RouteToApp.ClusterName, appServer)
-		if dialErr != nil {
-			if isReverseTunnelDownError(dialErr) {
-				t.c.log.Warnf("Failed to connect to application server %q: %v.", serverID, dialErr)
-				t.servers.Delete(serverID)
-				// Only goes for the next server if the error returned is a
-				// connection problem. Otherwise, stop iterating over the
-				// servers and return the error.
-				return true
-			}
+func (t *transport) DialContext(ctx context.Context, _, _ string) (conn net.Conn, err error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	servers := t.c.servers[:0]
```

Comment:
We never add to `t.c.servers`, right? Why not just zero out the head and keep the tail, if we're being fancy like this?